akka-actor-typed/src/main/scala/akka/actor/typed/internal/delivery/ProducerController.scala

```diff
@@ -0,0 +1,506 @@
+        (producer, initialState) match {
+          case (Some(p), Some(s)) => thenBecomeActive(p, c, s)
+          case (_, _)             => waitingForStart(context, producer, Some(c), durableQueue, initialState)(thenBecomeActive)
+        }
+      case start: Start[A] @unchecked =>
+        (consumerController, initialState) match {
+          case (Some(c), Some(s)) => thenBecomeActive(start.producer, c, s)
+          case (_, _) =>
+            waitingForStart(context, Some(start.producer), consumerController, durableQueue, initialState)(
+              thenBecomeActive)
+        }
+      case load: LoadStateReply[A] @unchecked =>
+        (producer, consumerController) match {
+          case (Some(p), Some(c)) => thenBecomeActive(p, c, load.state)
+          case (_, _) =>
+            waitingForStart(context, producer, consumerController, durableQueue, Some(load.state))(thenBecomeActive)
+        }
+      case LoadStateFailed(attempt) =>
+        // FIXME attempt counter, and give up
+        context.log.info("LoadState attempt [{}] failed, retrying.", attempt)
+        // retry
+        askLoadState(context, durableQueue, attempt + 1)
+        Behaviors.same
+    }
+  }
+
+  private def becomeActive[A: ClassTag](
+      producerId: String,
+      durableQueue: Option[ActorRef[DurableProducerQueue.Command[A]]],
+      state: State[A]): Behavior[InternalCommand] = {
+
+    Behaviors.setup { ctx =>
+      Behaviors.withTimers { timers =>
+        val msgAdapter: ActorRef[A] = ctx.messageAdapter(msg => Msg(msg))
+        val requested =
+          if (state.unconfirmed.isEmpty) {
+            state.producer ! RequestNext(producerId, 1L, 0L, msgAdapter, ctx.self)
+            true
+          } else {
+            ctx.self ! ResendFirst
+            false
+          }
+        new ProducerController[A](ctx, producerId, durableQueue, msgAdapter, timers)
+          .active(state.copy(requested = requested))
+      }
+    }
+  }
+
+}
+
+private class ProducerController[A: ClassTag](
+    ctx: ActorContext[ProducerController.InternalCommand],
+    producerId: String,
+    durableQueue: Option[ActorRef[DurableProducerQueue.Command[A]]],
+    msgAdapter: ActorRef[A],
+    timers: TimerScheduler[ProducerController.InternalCommand]) {
+  import ProducerController._
+  import ProducerController.Internal._
+  import ConsumerController.SequencedMessage
+  import DurableProducerQueue.StoreMessageSent
+  import DurableProducerQueue.StoreMessageSentAck
+  import DurableProducerQueue.StoreMessageConfirmed
+  import DurableProducerQueue.MessageSent
+
+  private implicit val askTimeout: Timeout = 3.seconds // FIXME config
+
+  private def active(s: State[A]): Behavior[InternalCommand] = {
+
+    def onMsg(m: A, newPendingReplies: Vector[(Long, ActorRef[Long])], ack: Boolean): Behavior[InternalCommand] = {
+      checkOnMsgRequestedState()
+      // FIXME adjust all logging, most should probably be debug
```

Comment:
General thought: perhaps we should also start using trace for high-volume/per-message-debg logging now that we have access to that through SLF4J?